File: synesis/connectors/chatgpt.py
from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path

from synesis.connectors.base import BaseConnector
from synesis.kb.types import ConversationMessage, RawConversation


class ChatGPTConnector(BaseConnector):
    name = "chatgpt"

    def __init__(self, config: dict):
        super().__init__(config)
        self.export_path = config.get("export_path") or ""

    def validate(self) -> bool:
        if not self.export_path:
            return False
        return Path(self.export_path).exists()

    def fetch(self, since: str | None = None) -> list[RawConversation]:
        if not self.export_path:
            return []

        since_dt = datetime.fromisoformat(since) if since else None
        conv_file = Path(self.export_path) / "conversations.json"
        if not conv_file.exists():
            return []

        conversations = json.loads(conv_file.read_text(encoding="utf-8"))
        results: list[RawConversation] = []

        for conv in conversations:
            conv_date = datetime.fromtimestamp(conv.get("update_time", 0))
            if since_dt and conv_date < since_dt:
                continue

            messages = []
            for node in sorted(
                conv.get("mapping", {}).values(),
                key=lambda n: (n.get("message", {}) or {}).get("create_time", 0) or 0,
            ):
                msg = node.get("message")
                if not msg or msg.get("author", {}).get("role") == "system":
                    continue

                parts = (msg.get("content", {}) or {}).get("parts", [])
                content = "\n".join(str(p) for p in parts if p).strip()
                if not content:
                    continue

                ts = msg.get("create_time")
                messages.append(
                    ConversationMessage(
                        role=msg["author"]["role"],
                        content=content,
                        timestamp=datetime.fromtimestamp(ts).isoformat() if ts else None,
                    )
                )

            if messages:
                title = conv.get("title", "untitled")
                slug = "".join(c if c.isalnum() else "-" for c in title.lower())[:60]
                results.append(
                    RawConversation(
                        source="chatgpt",
                        id=slug,
                        messages=messages,
                        timestamp=conv_date.isoformat(),
                        metadata={"title": title},
                    )
                )

        return results

File: src/connectors/chatgpt.ts
import fs from "fs/promises";
import path from "path";
import type { RawConversation, ConnectorConfig } from "../kb/types.js";
import { BaseConnector } from "./base.js";

interface ChatGPTExport {
  title: string;
  create_time: number;
  update_time: number;
  mapping: Record<string, {
    message?: {
      author: { role: string };
      content: { parts?: string[] };
      create_time?: number;
    };
  }>;
}

export class ChatGPTConnector extends BaseConnector {
  name = "chatgpt";
  private exportPath: string;

  constructor(config: ConnectorConfig) {
    super(config);
    this.exportPath = config.export_path as string || "";
  }

  async validate(): Promise<boolean> {
    if (!this.exportPath) return false;
    try {
      await fs.access(this.exportPath);
      return true;
    } catch {
      return false;
    }
  }

  async fetch(since?: Date): Promise<RawConversation[]> {
    if (!this.exportPath) return [];

    const raw = await fs.readFile(
      path.join(this.exportPath, "conversations.json"),
      "utf-8"
    );
    const conversations: ChatGPTExport[] = JSON.parse(raw);
    const results: RawConversation[] = [];

    for (const conv of conversations) {
      const convDate = new Date(conv.update_time * 1000);
      if (since && convDate < since) continue;

      const messages = Object.values(conv.mapping)
        .filter((m) => m.message && m.message.author.role !== "system")
        .sort((a, b) => (a.message?.create_time || 0) - (b.message?.create_time || 0))
        .map((m) => ({
          role: m.message!.author.role as "user" | "assistant",
          content: m.message!.content.parts?.join("\n") || "",
          timestamp: m.message!.create_time
            ? new Date(m.message!.create_time * 1000).toISOString()
            : undefined,
        }))
        .filter((m) => m.content.trim());

      if (messages.length > 0) {
        results.push({
          source: "chatgpt",
          id: conv.title.toLowerCase().replace(/[^a-z0-9]+/g, "-").slice(0, 60),
          messages,
          timestamp: convDate.toISOString(),
          metadata: { title: conv.title },
        });
      }
    }

    return results;
  }
}
